src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,395 @@
+#include <util/pointer_offset_size.h>
+#include <util/prefix.h>
+
+#include <linking/zero_initializer.h>
+
+#include <ansi-c/c_types.h>
+#include <ansi-c/string_constant.h>
+
+#include <goto-programs/goto_functions.h>
+
+#include "c_nondet_symbol_factory.h"
+
+class symbol_factoryt:public messaget
+{
+  code_blockt &init_code;
+  std::set<irep_idt> recursion_set;
+  bool assume_non_null;
+  symbol_tablet &symbol_table;
+  message_handlert &message_handler;
+  input_recordert &input_recorder;
+  namespacet ns;
+
+public:
+  symbol_factoryt(
+    code_blockt &_init_code,
+    bool _assume_non_null,
+    symbol_tablet &_symbol_table,
+    message_handlert &_message_handler,
+    input_recordert &input_recorder):
+      init_code(_init_code),
+      assume_non_null(_assume_non_null),
+      symbol_table(_symbol_table),
+      message_handler(_message_handler),
+      input_recorder(input_recorder),
+      ns(_symbol_table)
+  {}
+
+  exprt allocate_object(
+    const exprt &,
+    const typet &,
+    const source_locationt &,
+    const bool);
+
+  void gen_nondet_init(
+    const exprt &expr,
+    const source_locationt &loc);
+};
+
+/*******************************************************************\
+
+Function: symbol_factoryt::allocate_object
+
+ Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt symbol_factoryt::allocate_object(
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const source_locationt &loc,
+  const bool static_lifetime)
+{
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+  symbolt &aux_symbol=
+    c_declare_new_tmp_symbol(init_code, symbol_table, loc, allocate_type);
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  init_code.copy_to_operands(assign);
+
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  expr -
+  loc -
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  const exprt &expr,
+  const source_locationt &loc)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+
+    static size_t synthetic_constructor_count=0;
+
+    if(!assume_non_null)
+    {
```

Comment:
It would be nice to have a comment that sketches the generated code as pseudo-code.